blog/_posts/2017-05-29-tribulations-canbuildfrom.md

```diff
@@ -0,0 +1,193 @@
+
+[`CanBuildFrom`](/api/2.12.2/scala/collection/generic/CanBuildFrom.html) is probably the most
+infamous abstraction of the current collections. It is mainly criticised for making scary type
+signatures.
+
+Our ongoing [collections redesign](https://github.com/scala/collection-strawman) is an opportunity
+to  try alternative designs. This blogposts explains the (many!) problems solved by `CanBuildFrom`
+and the alternative solutions implemented in the new collections.
+
+## Transforming the elements of a collection
+
+It’s useful to think of `String` as a collection of `Char` elements: you can then use
+the common collection operations like `++`, `find`, etc. on `String` values.
+
+However the `map` method is challenging because this one
+transforms the `Char` elements into something that might or might not be `Char`s.
+Then, what should be the return type of the `map` method on `String` values? Ideally,
+we want to get back a `String` if we transform each `Char` into another `Char`, but we
+want to get some `Seq[B]` if we transform each `Char` into a different type `B`. And this
+is the way it currently works:
+
+~~~
+Welcome to Scala 2.12.2 (OpenJDK 64-Bit Server VM, Java 1.8.0_131).
+Type in expressions for evaluation. Or try :help.
+
+scala> "foo".map(c => c.toInt)
+res1: scala.collection.immutable.IndexedSeq[Int] = Vector(102, 111, 111)
+
+scala> "foo".map(c => c.toUpper)
+res2: String = FOO
+~~~
+
+This feature is not limited to the `map` method: `flatMap`, `collect`, `concat` and a few
+others also work the same. Moreover, `String` is not the only
+collection type that needs this feature: [`BitSet`](/api/2.12.2/index.html?search=bitset)
+and [`Map`](/api/2.12.2/index.html?search=map) are other examples.
+
+The current collections rely on `CanBuildFrom` to implement this feature. The `map`
+method is defined as follows:
+
+~~~ scala
+def map[B, That](f: Char => B)(implicit bf: CanBuildFrom[String, B, That]): That
+~~~
+
+When the implicit `CanBuildFrom` parameter is resolved it fixes the return type `That`.
+The resolution is driven by the actual `B` type: if `B` is `Char` then `That` is fixed
+to `String`, otherwise it is `immutable.IndexedSeq`.
+
+The drawback of this solution is that the type signature of the `map` method looks cryptic.
+
+In the new design we solve this problem by defining two overloads of the `map`
+method: one that handles `Char` to `Char` transformations, and one that handles other
+transformations. The type signatures of these `map` methods are straightforward:
+
+~~~ scala
+def map(f: Char => Char): String
+def map[B](f: Char => B): Seq[B]
+~~~
+
+Then, if you call `map` with a function that returns a `Char`, the first overload is
+selected and you get a `String`. Otherwise, the second overload is selected and you
+get a `Seq[B]`.
+
+Thus, we got rid of the cryptic method signatures while still supporting the feature
+of returning a different type of result according to the type of the transformation function.
```

Comment:
I think this point is not very important from the point of view of users. That’s more related to our internal implementation.